File: R/binsreg/tests/testthat/test-simulation-helpers.R
legacy_binsreg_pval <- function(num, denom, reps, tstat=NULL, side=NULL, alpha, lp=Inf) {
  tvec <- c()
  pval <- NA
  if (!is.null(tstat)) pval <- numeric(nrow(tstat))
  cval <- NA
  k <- ncol(num)

  for (i in seq_len(reps)) {
    eps <- matrix(rnorm(k, 0, 1), ncol = 1)
    tx <- (num %*% eps) / denom

    if (!is.null(side)) {
      if (side == "two") {
        if (is.infinite(lp)) tvec[i] <- max(abs(tx))
        else                 tvec[i] <- mean(abs(tx)^lp)^(1/lp)
      } else if (side == "left") {
        tvec[i] <- max(tx)
      } else if (side == "right") {
        tvec[i] <- min(tx)
      }
    }

    if (!is.null(tstat)) {
      for (j in seq_len(nrow(tstat))) {
        if (tstat[j, 2] == 1) {
          pval[j] <- pval[j] + (max(tx) >= tstat[j, 1])
        } else if (tstat[j, 2] == 2) {
          pval[j] <- pval[j] + (min(tx) <= tstat[j, 1])
        } else if (tstat[j, 2] == 3) {
          if (is.infinite(lp)) pval[j] <- pval[j] + (max(abs(tx)) >= tstat[j, 1])
          else                 pval[j] <- pval[j] + (mean(abs(tx)^lp)^(1/lp) >= tstat[j, 1])
        }
      }
    }
  }

  if (!is.null(tstat)) pval <- pval / reps
  if (!is.null(side)) cval <- quantile(tvec, alpha/100, na.rm=TRUE, names=FALSE, type=2)
  list(pval=pval, cval=cval)
}

legacy_binspwc_pval <- function(nummat1, nummat2, denom1, denom2, reps, tstat=NULL, testtype=NULL, lp=Inf, alpha=95) {
  pval <- 0
  tvec <- c()
  k1 <- ncol(nummat1)
  k2 <- ncol(nummat2)

  for (i in seq_len(reps)) {
    eps1 <- matrix(rnorm(k1, 0, 1), ncol = 1)
    eps2 <- matrix(rnorm(k2, 0, 1), ncol = 1)
    tx <- (nummat1 %*% eps1 - nummat2 %*% eps2) / sqrt(denom1^2+denom2^2)

    if (testtype == "left") {
      pval <- pval + (max(tx) >= tstat)
    } else if (testtype == "right") {
      pval <- pval + (min(tx) <= tstat)
    } else {
      if (is.infinite(lp)) pval <- pval + (max(abs(tx)) >= tstat)
      else                 pval <- pval + (mean(abs(tx)^lp)^(1/lp) >= tstat)
    }

    tvec[i] <- max(abs(tx))
  }

  list(
    pval=pval/reps,
    cval.cb=quantile(tvec, alpha/100, na.rm=TRUE, names=FALSE, type=2)
  )
}

test_that("binsreg.pval preserves the legacy simulation stream", {
  num <- matrix(c(
    0.4, -0.2, 0.7,
    1.1, 0.5, -0.3,
    -0.6, 0.9, 0.2,
    0.3, -0.8, 0.6
  ), nrow=4, byrow=TRUE)
  denom <- c(1.2, 0.9, 1.5, 1.1)
  tstat <- matrix(c(0.4, 1, -0.3, 2, 0.8, 3), ncol=2, byrow=TRUE)

  set.seed(60201)
  expected <- legacy_binsreg_pval(num, denom, 37, tstat=tstat, side="two", alpha=95, lp=Inf)
  set.seed(60201)
  observed <- binsreg:::binsreg.pval(num, denom, 37, tstat=tstat, side="two", alpha=95, lp=Inf)

  expect_equal(observed, expected, tolerance=1e-12)
})

test_that("binspwc.pval preserves the legacy simulation stream", {
  nummat1 <- matrix(c(
    0.1, 0.3, -0.5,
    0.7, -0.2, 0.4,
    -0.6, 0.8, 0.2,
    0.4, -0.1, 0.9
  ), nrow=4, byrow=TRUE)
  nummat2 <- matrix(c(
    -0.2, 0.6,
    0.5, -0.4,
    0.8, 0.1,
    -0.3, 0.7
  ), nrow=4, byrow=TRUE)
  denom1 <- c(1.0, 1.3, 0.8, 1.1)
  denom2 <- c(0.9, 0.7, 1.2, 1.4)

  set.seed(60202)
  expected <- legacy_binspwc_pval(nummat1, nummat2, denom1, denom2, 41, tstat=0.75, testtype="two-sided", alpha=95)
  set.seed(60202)
  observed <- binsreg:::binspwc.pval(nummat1, nummat2, denom1, denom2, 41, tstat=0.75, testtype="two-sided", alpha=95)

  expect_equal(observed, expected, tolerance=1e-12)
})

File: Python/binsreg/tests/test_simulation_helpers.py
from __future__ import annotations

import importlib.util

import pytest


np = pytest.importorskip("numpy")


def reference_binsreg_pval(num, denom, rep, tstat=None, side=None, alpha=95, lp=np.inf):
    tvec = np.empty(rep)
    pval = np.nan
    if tstat is not None:
        pval = np.zeros(tstat.shape[0])
    cval = np.nan
    k = num.shape[1]

    for i in range(rep):
        eps = np.random.normal(size=k).reshape(-1, 1)
        tx = np.matmul(num, eps) / denom.reshape(-1, 1)

        if side is not None:
            if side == "two":
                if np.isinf(lp):
                    tvec[i] = np.max(np.abs(tx))
                else:
                    tvec[i] = (np.mean(np.abs(tx) ** lp)) ** (1 / lp)
            elif side == "left":
                tvec[i] = np.max(tx)
            elif side == "right":
                tvec[i] = np.min(tx)

        if tstat is not None:
            for j in range(tstat.shape[0]):
                if tstat[j, 1] == 1:
                    pval[j] += np.max(tx) >= tstat[j, 0]
                elif tstat[j, 1] == 2:
                    pval[j] += np.min(tx) <= tstat[j, 0]
                elif tstat[j, 1] == 3:
                    if np.isinf(lp):
                        pval[j] += np.max(np.abs(tx)) >= tstat[j, 0]
                    else:
                        pval[j] += (np.mean(np.abs(tx) ** lp) ** (1 / lp)) >= tstat[j, 0]

    if tstat is not None:
        pval = pval / rep
    if side is not None:
        cval = np.quantile(tvec, alpha / 100)
    return pval, cval


def reference_binspwc_pval(nummat1, nummat2, denom1, denom2, rep, tstat=None, testtype=None, lp=np.inf):
    pval = 0
    k1 = nummat1.shape[1]
    k2 = nummat2.shape[1]
    denom1 = denom1.reshape(len(denom1), -1)
    denom2 = denom2.reshape(len(denom2), -1)

    for _ in range(rep):
        eps1 = np.random.normal(size=k1).reshape(-1, 1)
        eps2 = np.random.normal(size=k2).reshape(-1, 1)
        tx = (np.matmul(nummat1, eps1) - np.matmul(nummat2, eps2)) / np.sqrt(denom1**2 + denom2**2)

        if testtype == "left":
            pval += np.max(tx) >= tstat
        elif testtype == "right":
            pval += np.min(tx) <= tstat
        else:
            if not np.isfinite(lp):
                pval += np.max(np.abs(tx)) >= tstat
            else:
                pval += np.mean(np.abs(tx) ** lp) ** (1 / lp) >= tstat

    return pval / rep


def test_binsreg_pval_matches_reference_loop() -> None:
    if importlib.util.find_spec("binsreg") is None:
        pytest.skip("binsreg package is not importable")

    from binsreg.funs import binsreg_pval

    rng = np.random.default_rng(20240513)
    num = rng.normal(size=(17, 5))
    denom = 0.75 + rng.uniform(size=17)
    tstat = np.array([[0.5, 1], [-0.25, 2], [1.5, 3]], dtype=float)

    np.random.seed(98765)
    expected_pval, expected_cval = reference_binsreg_pval(
        num, denom, rep=200, tstat=tstat, side="two", alpha=95, lp=np.inf
    )
    np.random.seed(98765)
    actual_pval, actual_cval = binsreg_pval(
        num, denom, rep=200, tstat=tstat, side="two", alpha=95, lp=np.inf
    )

    np.testing.assert_allclose(actual_pval, expected_pval, rtol=0, atol=0)
    np.testing.assert_allclose(actual_cval, expected_cval, rtol=0, atol=1e-14)


def test_binspwc_pval_matches_reference_loop() -> None:
    if importlib.util.find_spec("binsreg") is None:
        pytest.skip("binsreg package is not importable")

    from binsreg.funs import binspwc_pval

    rng = np.random.default_rng(20240513)
    nummat1 = rng.normal(size=(13, 4))
    nummat2 = rng.normal(size=(13, 3))
    denom1 = 0.75 + rng.uniform(size=13)
    denom2 = 0.75 + rng.uniform(size=13)

    for testtype, tstat, lp in (
        ("left", 0.5, np.inf),
        ("right", -0.5, np.inf),
        ("two-sided", 1.25, np.inf),
        ("two-sided", 1.1, 2),
    ):
        np.random.seed(98765)
        expected = reference_binspwc_pval(
            nummat1, nummat2, denom1, denom2, rep=200, tstat=tstat, testtype=testtype, lp=lp
        )
        np.random.seed(98765)
        actual = binspwc_pval(
            nummat1, nummat2, denom1, denom2, rep=200, tstat=tstat, testtype=testtype, lp=lp
        )

        np.testing.assert_allclose(actual, expected, rtol=0, atol=0)


def test_fast_qreg_fit_matches_statsmodels() -> None:
    if importlib.util.find_spec("binsreg") is None:
        pytest.skip("binsreg package is not importable")

    sm = pytest.importorskip("statsmodels.api")
    from binsreg.funs import binsreg_qreg_fit

    rng = np.random.default_rng(20240513)
    x = np.column_stack((np.ones(80), rng.normal(size=(80, 4))))
    beta = np.array([0.25, -0.5, 0.75, 0.1, -0.2])
    y = x @ beta + rng.standard_t(df=5, size=80)

    actual = binsreg_qreg_fit(y, x, q=0.5)
    expected = sm.QuantReg(y, x).fit(q=0.5)

    assert actual.iterations == expected.iterations
    np.testing.assert_allclose(actual.params, expected.params, rtol=0, atol=1e-10)
    np.testing.assert_allclose(actual.fittedvalues, expected.fittedvalues, rtol=0, atol=1e-10)
    np.testing.assert_allclose(actual.cov_params(), expected.cov_params(), rtol=0, atol=1e-10)
